components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/graphql/GraphQLOperationAnalyzer.java

```diff
@@ -0,0 +1,556 @@
+            FieldComplexityEnvironment fieldComplexityEnvironment = convertEnv(queryVisitorFieldEnvironment);
+            return fieldComplexityCalculator.calculate(fieldComplexityEnvironment, childsComplexity);
+        }
+    }
+
+    private static FieldComplexityEnvironment convertEnv(QueryVisitorFieldEnvironment queryVisitorFieldEnvironment) {
+        FieldComplexityEnvironment parentEnv = null;
+        if (queryVisitorFieldEnvironment.getParentEnvironment() != null) {
+            parentEnv = convertEnv(queryVisitorFieldEnvironment.getParentEnvironment());
+        }
+
+        return new FieldComplexityEnvironment(queryVisitorFieldEnvironment.getField(),
+                queryVisitorFieldEnvironment.getFieldDefinition(),
+                queryVisitorFieldEnvironment.getFieldsContainer(),
+                queryVisitorFieldEnvironment.getArguments(),
+                parentEnv);
+    }
+
+    /**
+     * @param objectType
+     * @param operationName
+     * @return
+     */
+    public static String getOperationReturnType(GraphQLObjectType objectType, String operationName) {
+        String returnType = null;
+        for (GraphQLFieldDefinition fieldDefinition : objectType.getFieldDefinitions()) {
+            if (fieldDefinition.getName().equals(operationName)) {
+                if (fieldDefinition.getType() instanceof GraphQLInterfaceType) {
+                    returnType = ((GraphQLInterfaceType) fieldDefinition.getType()).getName();
+                } else if (fieldDefinition.getType() instanceof GraphQLObjectType) {
+                    returnType = ((GraphQLObjectType) fieldDefinition.getType()).getName();
+                } else if (fieldDefinition.getType() instanceof GraphQLList){
+                    returnType = ((GraphQLObjectType)((GraphQLList)fieldDefinition.getType()).getOriginalWrappedType()).getName();
+                }
+            }
+        }
+        return returnType;
+    }
+
+    /**
+     * @param operation
+     * @param typeDefinitionRegistry
+     * @return
+     */
+    public static Map<String, Object> getTypeList(OperationDefinition operation,
+                                                  TypeDefinitionRegistry typeDefinitionRegistry) {
+        Map<String, Object> definedTypes = new HashMap<>();
+        String type = operation.getOperation().toString();
+        Map<String, Integer> fieldUsage = analyzePayload(operation);
+        Map<String, List<String>> result = processFieldUsage(fieldUsage);
+        definedTypes = GraphQLOperationAnalyzer.getSchemaUsage(result, definedTypes, typeDefinitionRegistry, type);
+        return definedTypes;
+    }
+
+    private static Map<String, Integer> analyzePayload(OperationDefinition operation) {
+        Map<String, Integer> fieldUsage = new HashMap<>();
+        List<Selection> subDictionary = operation.getSelectionSet().getSelections();
+
+        for (Selection selection : subDictionary) {
+            Field levelField = (Field) selection;
```

Comment:
Is this always a Field? If yes, why we have checked instanceof in (https://github.com/wso2/carbon-apimgt/pull/12441/files#diff-5f773ffd97be42a867859fa6ce432df90e6833588cf7c1fd3e54a6f26cb24e7aR79)

Reply:
Fixed via https://github.com/wso2/carbon-apimgt/pull/12441/commits/226c49a261240dc5305f96d69aadb6f86edfae84